### PR TITLE
winch(x64): Initial implementation for function calls

### PR DESCRIPTION
**Pull request by saulecabrera**
This change adds the main building blocks for calling locally defined
functions. Support for function imports will be added iteratively after this
change lands and once trampolines are supported. Which is also when 
we plan to introduce support for the `Wasmtime*` calling conventions.

To support function calls, this change introduces the following functionality to
the MacroAssembler:

* `pop` to pop the machine stack into a given register, which in the case of
this change, translates to the x64 pop instruction.

* `call` to a emit a call to locally defined functions.

* `address_from_sp` to construct memory addresses with the SP as a base.

* `free_stack` to emit the necessary instrunctions to claim stack space.

The heavy lifting of setting up and emitting the function call is done through
the implementation of `FnCall`.

### CODE REVIEW

**Comment by saulecabrera** [comment]
Oh It seems I messed up the branches and that's why https://github.com/bytecodealliance/wasmtime/pull/6067/commits/03637c95fa75767a9c4cfedd638cb1f45016053d which is unrelated to this change appears. I'll fix that.

File: winch/codegen/src/codegen/context.rs [hunk 1]
```diff
@@ -60,58 +61,66 @@ impl<'a> CodeGenContext<'a> {
...
-    /// Checks if the stack top contains the given register. The register
-    /// gets allocated otherwise, potentially causing a spill.
-    /// Once the requested register is allocated, the value at the top of the stack
-    /// gets loaded into the register.
-    pub fn pop_to_named_reg<M: MacroAssembler>(
+    /// Loads the stack top value into the next available register, if
+    /// it isn't already one; spilling if there are no registers
+    /// available.  Optionally the caller may specify a specific
+    /// destination register.
+    pub fn pop_to_reg<M: MacroAssembler>(
```
Comment (saulecabrera) [inline winch/codegen/src/codegen/context.rs:68]: The work by `pop_to_reg` and `pop_to_named_reg` was very similar, so I realized I could remove some repetition here through this refactor.

File: winch/codegen/src/codegen/call.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        masm: &mut M,
+    ) -> Self {
+        let stack = &context.stack;
+        let arg_stack_space = callee_sig.stack_bytes;
+        let callee_params = &callee_sig.params;
+        let sp_offset_at_callsite = masm.sp_offset();
+
+        let spilled_regs = match callee_params.len() {
+            0 => {
+                let _ = context.spill_regs_in(masm, ..);
```
Comment (cfallin) [inline winch/codegen/src/codegen/call.rs:53]: Does this need to be a special case? It seems that if instead the `spill_regs_in` method accepted an empty range, this case would work with the code below...

File: winch/codegen/src/codegen/call.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
+use super::CodeGenContext;
```
Comment (cfallin) [inline winch/codegen/src/codegen/call.rs:1]: A top-level module comment describing the calling convention would be really helpful here! In particular, which registers are clobbered or saved, and where arguments are passed.

File: winch/filetests/filetests/x64/call/params.wat [hunk 4]
```diff
@@ -0,0 +1,99 @@
+;;! target = "x86_64"
```
Comment (cfallin) [inline winch/filetests/filetests/x64/call/params.wat:1]: It might be good to have a test initially that exercises the spilling behavior (here we can't because `$add` is called in tail position). Perhaps call it multiple times, with some values on the stack that live "under" the call?

**Review by cfallin — APPROVED** [review]
Looks generally great, thanks! Just some clarification-comment thoughts and possibly another test case.

**Comment by saulecabrera** [comment]
@cfallin thanks for the review and sorry it took me a bit to get back to this. 

Good call on introducing a more complex test -- I was able to uncover a bug regarding addresses in the spill behaviour. I've introduced a fix and an explanation in https://github.com/bytecodealliance/wasmtime/pull/6067/commits/5e058d21165477cd4ae1d1257c7f7f218d604f50. Given that I introduced some new changes, I'd appreciate if you could take another look. Thanks!

File: winch/codegen/src/codegen/call.rs (hunk 2 above)
Comment (saulecabrera) [reply winch/codegen/src/codegen/call.rs]: I was not able to find a better way to do this. In this case we still want to spill everything, to save any live registers, but we are no interested in the count returned by `spill_regs_and_count_memory_in`, since they are not part of the function arguments so we can't claim that memory after the function call, so that's why I'm handling `0` in this way.

File: winch/codegen/src/codegen/call.rs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        let spilled_regs = match callee_params.len() {
+            0 => {
+                let _ = context.spill_regs_in(masm, ..);
+                0
+            }
+            _ => {
+                assert!(stack.len() >= callee_params.len());
+                let partition = stack.len() - callee_params.len();
+                let _ = context.spill_regs_in(masm, 0..partition);
+                context.spill_regs_in(masm, partition..)
```
[the remark this answers is not in the document]
Comment (saulecabrera) [reply winch/codegen/src/codegen/call.rs]: Fixed!

File: winch/filetests/filetests/x64/call/params.wat (hunk 4 above)
Comment (saulecabrera) [reply winch/filetests/filetests/x64/call/params.wat:1]: Fixed! Thanks.

File: winch/codegen/src/codegen/call.rs (hunk 3 above)
Comment (saulecabrera) [reply winch/codegen/src/codegen/call.rs:3]: Agreed! Added the documentation in the abi module, and added a reference to it from here.

File: winch/codegen/src/codegen/call.rs [hunk 6]
```diff
@@ -11,15 +13,24 @@ pub(crate) struct FnCall<'a> {
...
-    pub spill_stack_space: u32,
+    spill_stack_space: u32,
+    /// The amount of memory in bytes consumed by the call. That is,
+    /// any memory entries that will be used as params and that exist
+    /// before any spills generated by saving live registers are
+    /// counted as memory entries that will be consumed by the call,
+    /// and thus we track this space to claim it once the function
+    /// call has ended. This field is analogous to
+    /// `spill_stack_space`, but it's kept separate for clarity and
+    /// debugging purposes.
```
Comment (cfallin) [inline winch/codegen/src/codegen/call.rs:30]: I definitely appreciate the docs here, but a few things are slightly unclear to me at least:

- The terminology of "memory entries" is confusing -- does this just mean "values on the stack"?
- The "That is, any memory entries ... " sentence is really hard to parse. I *think* what this is saying is: values already on the stack that are args?
- "Claim" means "reclaim" here?

Is it the case that `memory_consumed + spill_stack_space == total_arg_stack_space`? If so, could we use that alone to pop args off the stack? Fundamentally it seems like we shouldn't have to know where the args originally came from when we're cleaning up.

File: winch/codegen/src/codegen/context.rs [hunk 7]
```diff
@@ -182,31 +182,35 @@ impl<'a> CodeGenContext<'a> {
     /// range defined by the caller.  This is a specialization of the
     /// spill function; made available for cases in which spilling
     /// locals is not required, like for example for function calls in
-    /// which locals are not reachable by the callee.
+    /// which locals are not reachable by the callee.  It also tracks
+    /// down the amount of memory entries in the given range.
     ///
-    /// Returns the amount of registers spilled for the given range.
-    pub fn spill_regs_in<M, R>(&mut self, masm: &mut M, range: R) -> u32
+    /// Returns the amount of registers spilled and the amount of
```
Comment (cfallin) [inline winch/codegen/src/codegen/context.rs:188]: "Amount of registers" -> "number of registers"? (register is a count-noun so the "amount of" reads ambiguously to me: in bytes, or whole registers, or ...)

File: winch/codegen/src/codegen/call.rs [hunk 8]
```diff
@@ -87,9 +125,9 @@ impl<'a> FnCall<'a> {
         masm.reserve_stack(self.total_arg_stack_space);
         self.assign_args(context, masm, <A as ABI>::scratch_reg());
         masm.call(callee);
-        masm.free_stack(self.spill_stack_space + self.total_arg_stack_space);
+        masm.free_stack(self.spill_stack_space + self.total_arg_stack_space + self.memory_consumed);
```
Comment (cfallin) [inline winch/codegen/src/codegen/call.rs:128]: I'm a little confused here: shouldn't we be popping just `total_arg_stack_space`, without also popping `memory_consumed` and `spill_stack_space`? It seems like this is double-counting, unless I'm misunderstanding what the above amounts mean.

(Labeling portions of the stack in the figure might help clarify, too, or else noting the state of hte stack at points in the sequence above)

File: winch/codegen/src/codegen/call.rs (hunk 6 above)
Comment (saulecabrera) [reply winch/codegen/src/codegen/call.rs:30]: > Is it the case that memory_consumed + spill_stack_space == total_arg_stack_space?

I see where you're coming from and I agree that it's a bit confusing, I think I can make it so that it's more clear and so that we have to deal with less terms.

But to give a bit more context:

* total_arg_stack_space is the total space that needs to be explicitly allocated at the callsite, and covers: the stack space needed for arguments plus any alignment needed for the caller's frame. 

* memory_consumed, represents the memory used by any `Memory` entries existing in the value stack at the moment of assigning those entries to the function arguments (in the case of a register argument, that memory will be loaded to the register used as argument, and in the case of a stack argument, that memory will be loaded to the scratch register and then stored to the particular argument slot). We could emit a `pop` instruction when doing this assignment and will potentially simplify things here in terms of what we are tracking, but I think it's more efficient to track the amount of memory used and reclaim it in the end.

* spill_stack_space is the space created to save any live registers.

Does that clarify things to you? 

As I write this, I believe we can simplify this by tracking all the memory used in a single field. Initially I thought that separating where things come from would make the code more understandable and easier to reason about, but maybe the opposite is better.

Comment (saulecabrera) [reply winch/codegen/src/codegen/call.rs]: I've simplified how we track these values in https://github.com/bytecodealliance/wasmtime/pull/6067/commits/2e5372c8ffb430e75442ef33cc5cd7a3b408800e and also added more extensive documentation. Hopefully this clarifies the code.

File: winch/codegen/src/codegen/call.rs (hunk 8 above)
Comment (saulecabrera) [reply winch/codegen/src/codegen/call.rs]: Agreed, I've added a figure of how the stack looks like at each point in time in https://github.com/bytecodealliance/wasmtime/commit/2e5372c8ffb430e75442ef33cc5cd7a3b408800e, hopefully this helps clarify!

**Review by cfallin — APPROVED** [review]
This is clearer now with the updated comments and after our discussion today -- thanks very much for the patience! LGTM.
